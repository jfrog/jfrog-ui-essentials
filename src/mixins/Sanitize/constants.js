const BASE64_TEST_PNG_IMAGE = `iVBORw0KGgoAAAANSUhEUgAAADcAAAA3CAAAAACNsI2aAAAACXBIWXMAAAB5AAAAeQBPsriEAAAB6ElEQVR42rVWO46EMAzNadAcY3vaOQMXoXcXKZehS8NpqNxamw8JxDYra1Zjhgge9jhx/By7bYvtl4Y8Qn+tEjty6WxuQ0KkfOM5wJEeEkT1bsigU+xGQV+QfZ2ned0LAkLnyQ4XV2XB/k+jXdTs8Mc1+UlvQehEt5Fit7hLFsUfqfOk3d1lJ9VO+qN1sFvJm+IScB7s3uo8ZVzC8RrsXjIuqp2n0d+sxFNbHxCw9cF34yn2L5jyJWndIprzRfqLpvw0+6PCh1fjgxpP5NL4VzlYEa6zOYDgzyvk0cMbykMek6THipSXAD5/BKh8H/3JGZTxPgM9Px9WDL0CkM1ORJie48nsWAXQ8kW1YxlknKfIWJs/EBXgoZ6Jf2KMNMYz4FgBJjTGkxR/H67vm/H8eP9ShlyRqfli24c0svy0zLNXgOkNtQJEle/P/MPOv8T3TGZIZIbO7sL7BMON74nkuQqUj4XvnMvwiNCBjO+yev2NVDtZLeX5rvD9lu0zauxW+a6dBvJ8H5Gyfzz3wIBkO57rYECyHeeWF+xW+YcT47Jkdzi4TpT+lPNdIv9Z34fxNOxf0PhO91yw5MuMen56AxLPOtG7W9T63SCQ2k9Uol1so3bVnrog2JTyU57n1bb37n3s5s8Of5RfsaTdSlfuyUAAAAA8dEVYdGNvbW1lbnQAIEltYWdlIGdlbmVyYXRlZCBieSBHTlUgR2hvc3RzY3JpcHQgKGRldmljZT1wbm1yYXcpCvqLFvMAAABKdEVYdHNpZ25hdHVyZQA4NWUxYWU0YTJmYmE3OGVlZDRmZDhmMGFjZjIzNzYwOWU4NGY1NDk2Y2RlMjBiNWQ3NmM5Y2JjMjk4YzRhZWJjJecJ2gAAAABJRU5ErkJggg==`
const TEST_SIMPLE_TAGS = [
    { tag: 'span', isSelfClosing: false },
    { tag: 'h1', isSelfClosing: false, allowedAttributes: ['align="left"'] },
    { tag: 'h2', isSelfClosing: false, allowedAttributes: ['align="left"']  },
    { tag: 'h3', isSelfClosing: false, allowedAttributes: ['align="left"']  },
    { tag: 'h4', isSelfClosing: false, allowedAttributes: ['align="left"']  },
    { tag: 'h5', isSelfClosing: false, allowedAttributes: ['align="left"']  },
    { tag: 'h6', isSelfClosing: false, allowedAttributes: ['align="left"']  },
    { tag: 'blockquote', isSelfClosing: false },
    { tag: 'p', isSelfClosing: false, allowedAttributes: [
            'align="left"',
            'cite="http://www.mycites.com"',
            'cite="https://www.mycites.com"',
            'cite="ftp://10.20.30:4050/cites"'
        ]
    },
    {
        tag: 'a', isSelfClosing: false, allowedAttributes: [
            'class="moshe"',
            'href="mailto:my@example.com"',
            'href="tel:123456789"',
        ]
    },
    { tag: 'ul', isSelfClosing: false },
    { tag: 'ol', isSelfClosing: false },
    { tag: 'nl', isSelfClosing: false },
    { tag: 'li', isSelfClosing: false },
    { tag: 'li', isSelfClosing: false },
    { tag: 'b', isSelfClosing: false },
    { tag: 'i', isSelfClosing: false },
    { tag: 'strong', isSelfClosing: false },
    { tag: 'em', isSelfClosing: false },
    { tag: 'strike', isSelfClosing: false },
    { tag: 'code', isSelfClosing: false },
    { tag: 'hr', isSelfClosing: true },
    { tag: 'br', isSelfClosing: true },
    { tag: 'div', isSelfClosing: false, allowedAttributes: ['class="moshe"', 'id="moshe"'] },
    { tag: 'pre', isSelfClosing: false },
    { tag: 'table', isSelfClosing: false },
    {
        tag: 'img', isSelfClosing: true, allowedAttributes: [
            'src="/my/relative/img/src"',
            'src="http://www.myphotos.com"',
            'src="https://www.myphotos.com"',
            'src="ftp://10.20.30:4050/photos"',
            'width="100px"'
        ]
    },
    { tag: 'area', isSelfClosing: true },
    // These tags must be as child of specified parents (in ascending order)
    { tag: 'thead', isSelfClosing: false, parentTags: ['table']},
    { tag: 'caption', isSelfClosing: false, parentTags: ['table']},
    { tag: 'tbody', isSelfClosing: false, parentTags: ['table']},
    { tag: 'tr', isSelfClosing: false, parentTags: ['tbody', 'table']},
    {
        tag: 'td', isSelfClosing: false, parentTags: ['tr', 'tbody', 'table'],
        allowedAttributes: [
            'align="left"', 'valign="bottom"'
        ]
    },
];
const TEST_EVENTS = ['hover', 'click', 'error'];

export { TEST_EVENTS, TEST_SIMPLE_TAGS, BASE64_TEST_PNG_IMAGE };
