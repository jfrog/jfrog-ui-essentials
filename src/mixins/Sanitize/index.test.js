import sanitizeMixin from './index';
import { BASE64_TEST_PNG_IMAGE, TEST_EVENTS, TEST_SIMPLE_TAGS } from './constants';
import { createLocalVue, shallowMount } from '@vue/test-utils';
import Vue from 'vue';

const HtmlSanitizeTestComponent = {
    template: `
    <div>
        <div class="safe" v-html="$sanitize(dynamicContentUnsafe,config)"></div>
        <div class="unsafe" v-html="dynamicContentUnsafe"></div>
    </div>
    `,
    name: 'HtmlSanitizeTestComponent',
    mixins: [sanitizeMixin],
    props: {
        dynamicContentUnsafe: String,
        config: Object,
    },
};
const URLSanitizeTestComponent = {
    template: `
    <div>
        <a class="safe-anchor" :href="$sanitizeUrl(dynamicUrl)"></a>
        <a class="unsafe-anchor" :href="dynamicUrl"></a>
    </div>
    `,
    name: 'URLSanitizeTestComponent',
    mixins: [sanitizeMixin],
    props: {
        dynamicUrl: String,
    },
};

const getElements = (wrapper) => {
    const unsafe = wrapper.find('.unsafe');
    const safe = wrapper.find('.safe');
    return { unsafe, safe };
};

const getHtml = (wrapper) => {
    const { unsafe, safe } = getElements(wrapper);
    const unsafeHtml = unsafe.html();
    const safeHtml = safe.html();
    return { unsafeHtml, safeHtml };
};


const allTagEventCombinations = [];
TEST_SIMPLE_TAGS.forEach(({ tag, isSelfClosing, parentTags = [] }) => {
    TEST_EVENTS.forEach(event => {
        allTagEventCombinations.push([tag, event, isSelfClosing, parentTags]);
    });
});

describe('Sanitize Html Mixin', () => {
    describe('Special tags test suit', () => {
        let wrapper;
        beforeEach(() => {
            const localVue = createLocalVue();
            wrapper = shallowMount(HtmlSanitizeTestComponent, {
                localVue,
            });
        });

        it('rendering', () => {
            expect(wrapper.find(HtmlSanitizeTestComponent).exists()).toBe(true);
        });

        it('Should escape script tag with onclick XSS', async () => {
            const xss = `<script onclick="alert('xss')"></script>`;
            wrapper.setProps({ dynamicContentUnsafe: xss });
            await Vue.nextTick();
            const { unsafeHtml, safeHtml } = getHtml(wrapper);
            expect(unsafeHtml).toContain(xss);
            expect(safeHtml).toContain('&lt;script&gt;&lt;/script&gt;');
            expect(wrapper.element).toMatchSnapshot();
        });

        it('Should escape script tag with inner js XSS', async () => {
            const xss = `<script>alert('xss')</script>`;
            wrapper.setProps({ dynamicContentUnsafe: xss });
            await Vue.nextTick();
            const { unsafeHtml, safeHtml } = getHtml(wrapper);
            expect(unsafeHtml).toContain(`<script>`);
            expect(unsafeHtml).toContain(`alert('xss')`);
            expect(safeHtml).toContain(`&lt;script&gt;alert('xss')&lt;/script&gt;`);
            expect(wrapper.element).toMatchSnapshot();
        });

        it('Should escape object tag with data64 XSS', async () => {
            // PHNjcmlwdD5hbGVydCgiWFNTIik7PC9zY3JpcHQ is '<script onclick="alert('xss')"></script>'
            const xss = `<object data="data:text/html;base64,PHNjcmlwdD5hbGVydCgiWFNTIik7PC9zY3JpcHQ+Cg=="></object>`;
            wrapper.setProps({ dynamicContentUnsafe: xss });
            await Vue.nextTick();
            const { unsafeHtml, safeHtml } = getHtml(wrapper);
            expect(unsafeHtml).toContain(xss);
            expect(safeHtml).toContain('&lt;object&gt;&lt;/object&gt;');
            expect(wrapper.element).toMatchSnapshot();
        });

        it('Should sanitize img tag with onerror XSS', async () => {
            const xss = `<img src="x" onerror="alert('xss');">`;
            wrapper.setProps({ dynamicContentUnsafe: xss });
            await Vue.nextTick();
            const { unsafeHtml, safeHtml } = getHtml(wrapper);
            expect(unsafeHtml).toContain(xss);
            expect(safeHtml).toContain('<img src="x">');
            expect(wrapper.element).toMatchSnapshot();
        });

        it('Should not sanitize custom configured tags', async () => {
            const mockHtml = `<ul><li><input checked="" disabled="" type="checkbox" /> foo<ul><li><input disabled="" type="checkbox" /> bar</li><li><input checked="" disabled="" type="checkbox" /> baz</li></ul><li><input disabled="" type="checkbox" /> bim</li></ul>`;
            const { DEFAULT_CONFIG } = sanitizeMixin;
            const releaseNotesSanitizerConfig = JSON.parse(JSON.stringify(DEFAULT_CONFIG));
            releaseNotesSanitizerConfig.allowedTags.push('input');
            releaseNotesSanitizerConfig.allowedAttributes.input = ['checked', 'disabled', 'type'];
            wrapper.setProps({ dynamicContentUnsafe: mockHtml, config: releaseNotesSanitizerConfig });
            await Vue.nextTick();
            expect(wrapper.element).toMatchSnapshot();
        });
        it('Should sanitize & render img with base64 src', async () => {
            const xss = `<img src="data:image/png;base64,${BASE64_TEST_PNG_IMAGE}" onerror="alert('xss');">`;
            wrapper.setProps({ dynamicContentUnsafe: xss });
            await Vue.nextTick();
            const { unsafeHtml, safeHtml } = getHtml(wrapper);
            expect(unsafeHtml).toContain(xss);
            expect(safeHtml).toContain(`<img src="data:image/png;base64,${BASE64_TEST_PNG_IMAGE}">`);
            expect(wrapper.element).toMatchSnapshot();
        });

        it('Should sanitize & keep href attribute in anchor tag', async () => {
            const xss = `<a href="http://google.com" onerror="alert('xss');"></a>`;
            wrapper.setProps({ dynamicContentUnsafe: xss });
            await Vue.nextTick();
            const { unsafeHtml, safeHtml } = getHtml(wrapper);
            expect(unsafeHtml).toContain(xss);
            expect(safeHtml).toContain('<a href="http://google.com"></a>');
            expect(wrapper.element).toMatchSnapshot();
        });
        it('Should sanitize javascript scheme from anchor tag', async () => {
            const xss = `<a href="javascript:alert('xss');"></a>`;
            wrapper.setProps({ dynamicContentUnsafe: xss });
            await Vue.nextTick();
            const { unsafeHtml, safeHtml } = getHtml(wrapper);
            expect(unsafeHtml).toContain(xss);
            expect(safeHtml).toContain('<a></a>');
            expect(wrapper.element).toMatchSnapshot();
        });
    });

    describe('Simple tags data driven test suit', () => {
        let wrapper;

        beforeEach(() => {
            const localVue = createLocalVue();
            wrapper = shallowMount(HtmlSanitizeTestComponent, {
                localVue,
            });
        });

        test.each(allTagEventCombinations)(`Should sanitize %s tag with %s event XSS`,
            async (tag, event, isSelfClosing, parentTags) => {
                const closingTag = isSelfClosing ? '' : `</${tag}>`;
                let xss = `<${tag} on${event}="eventHandler">${closingTag}`;
                if (parentTags && parentTags.length) {
                    parentTags.forEach(parentTag => {
                        xss = `<${parentTag}>${xss}</${parentTag}>`;
                    });
                }
                wrapper.setProps({ dynamicContentUnsafe: xss });
                await Vue.nextTick();
                const { unsafeHtml, safeHtml } = getHtml(wrapper);
                expect(unsafeHtml).toContain(`${tag}`);
                expect(unsafeHtml).toContain(`on${event}="eventHandler"`);
                expect(safeHtml).toContain(`<${tag}>${closingTag}`);
                expect(!safeHtml.includes(`on${event}="eventHandler"`));
            });
    });
    describe('Sanitize URL', () => {
        let wrapper;

        beforeEach(() => {
            const localVue = createLocalVue();
            wrapper = shallowMount(URLSanitizeTestComponent, {
                localVue,
            });
        });
        it('Should sanitize url with javascript in href', async () => {
            const xss = `javascript:alert('xss')`;
            wrapper.setProps({ dynamicUrl: xss });
            await Vue.nextTick();
            const safeUrl = wrapper.find('.safe-anchor').attributes('href');
            const unsafeUrl = wrapper.find('.unsafe-anchor').attributes('href');
            expect(unsafeUrl).toContain(xss);
            expect(safeUrl).toContain('about:blank');
            expect(wrapper.element).toMatchSnapshot();
        });
        it('Should sanitize url with encoded javascript in href', async () => {
            const xss = decodeURIComponent(`JaVaScRiP%0at:alert('xss')`);
            console.log(xss);
            wrapper.setProps({ dynamicUrl: xss });
            await Vue.nextTick();
            const safeUrl = wrapper.find('.safe-anchor').attributes('href');
            const unsafeUrl = wrapper.find('.unsafe-anchor').attributes('href');
            expect(unsafeUrl).toContain(xss);
            expect(safeUrl).toContain('about:blank');
            expect(wrapper.element).toMatchSnapshot();
        });
        it('Should not change a valid url', async () => {
            const noXss = 'https://example.com';
            wrapper.setProps({ dynamicUrl: noXss });
            await Vue.nextTick();
            const safeUrl = wrapper.find('.safe-anchor').attributes('href');
            const unsafeUrl = wrapper.find('.unsafe-anchor').attributes('href');
            expect(unsafeUrl).toContain(noXss);
            expect(safeUrl).toContain(noXss);
            expect(wrapper.element).toMatchSnapshot();
        });
        it('Should not change a valid url without https', async () => {
            const noXss = 'example.com';
            wrapper.setProps({ dynamicUrl: noXss });
            await Vue.nextTick();
            const safeUrl = wrapper.find('.safe-anchor').attributes('href');
            const unsafeUrl = wrapper.find('.unsafe-anchor').attributes('href');
            expect(unsafeUrl).toContain(noXss);
            expect(safeUrl).toContain(noXss);
            expect(wrapper.element).toMatchSnapshot();
        });
    });
});
