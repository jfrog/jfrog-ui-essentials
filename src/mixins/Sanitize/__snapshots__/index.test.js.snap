// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sanitize Html Mixin Sanitize URL Should not change a valid url 1`] = `
<div>
  <a
    class="safe-anchor"
    href="https://example.com"
  />
   
  <a
    class="unsafe-anchor"
    href="https://example.com"
  />
</div>
`;

exports[`Sanitize Html Mixin Sanitize URL Should not change a valid url without https 1`] = `
<div>
  <a
    class="safe-anchor"
    href="example.com"
  />
   
  <a
    class="unsafe-anchor"
    href="example.com"
  />
</div>
`;

exports[`Sanitize Html Mixin Sanitize URL Should sanitize url with encoded javascript in href 1`] = `
<div>
  <a
    class="safe-anchor"
    href="about:blank"
  />
   
  <a
    class="unsafe-anchor"
    href="JaVaScRiP
t:alert('xss')"
  />
</div>
`;

exports[`Sanitize Html Mixin Sanitize URL Should sanitize url with javascript in href 1`] = `
<div>
  <a
    class="safe-anchor"
    href="about:blank"
  />
   
  <a
    class="unsafe-anchor"
    href="javascript:alert('xss')"
  />
</div>
`;

exports[`Sanitize Html Mixin Special tags test suit Should escape object tag with data64 XSS 1`] = `
<div>
  <div
    class="safe"
  />
   
  <div
    class="unsafe"
  >
    <object
      data="data:text/html;base64,PHNjcmlwdD5hbGVydCgiWFNTIik7PC9zY3JpcHQ+Cg=="
    />
  </div>
</div>
`;

exports[`Sanitize Html Mixin Special tags test suit Should escape script tag with inner js XSS 1`] = `
<div>
  <div
    class="safe"
  />
   
  <div
    class="unsafe"
  >
    <script>
      alert('xss')
    </script>
  </div>
</div>
`;

exports[`Sanitize Html Mixin Special tags test suit Should escape script tag with onclick XSS 1`] = `
<div>
  <div
    class="safe"
  />
   
  <div
    class="unsafe"
  >
    <script
      onclick="alert('xss')"
    />
  </div>
</div>
`;

exports[`Sanitize Html Mixin Special tags test suit Should not sanitize custom configured tags 1`] = `
<div>
  <div
    class="safe"
  >
    <ul>
      <li>
        <input
          checked=""
          disabled=""
          type="checkbox"
        />
         foo
        <ul>
          <li>
            <input
              disabled=""
              type="checkbox"
            />
             bar
          </li>
          <li>
            <input
              checked=""
              disabled=""
              type="checkbox"
            />
             baz
          </li>
        </ul>
      </li>
      <li>
        <input
          disabled=""
          type="checkbox"
        />
         bim
      </li>
    </ul>
  </div>
   
  <div
    class="unsafe"
  >
    <ul>
      <li>
        <input
          checked=""
          disabled=""
          type="checkbox"
        />
         foo
        <ul>
          <li>
            <input
              disabled=""
              type="checkbox"
            />
             bar
          </li>
          <li>
            <input
              checked=""
              disabled=""
              type="checkbox"
            />
             baz
          </li>
        </ul>
      </li>
      <li>
        <input
          disabled=""
          type="checkbox"
        />
         bim
      </li>
    </ul>
  </div>
</div>
`;

exports[`Sanitize Html Mixin Special tags test suit Should sanitize & keep href attribute in anchor tag 1`] = `
<div>
  <div
    class="safe"
  >
    <a
      href="http://google.com"
    />
  </div>
   
  <div
    class="unsafe"
  >
    <a
      href="http://google.com"
      onerror="alert('xss');"
    />
  </div>
</div>
`;

exports[`Sanitize Html Mixin Special tags test suit Should sanitize & render img with base64 src 1`] = `
<div>
  <div
    class="safe"
  >
    <img
      src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAADcAAAA3CAAAAACNsI2aAAAACXBIWXMAAAB5AAAAeQBPsriEAAAB6ElEQVR42rVWO46EMAzNadAcY3vaOQMXoXcXKZehS8NpqNxamw8JxDYra1Zjhgge9jhx/By7bYvtl4Y8Qn+tEjty6WxuQ0KkfOM5wJEeEkT1bsigU+xGQV+QfZ2ned0LAkLnyQ4XV2XB/k+jXdTs8Mc1+UlvQehEt5Fit7hLFsUfqfOk3d1lJ9VO+qN1sFvJm+IScB7s3uo8ZVzC8RrsXjIuqp2n0d+sxFNbHxCw9cF34yn2L5jyJWndIprzRfqLpvw0+6PCh1fjgxpP5NL4VzlYEa6zOYDgzyvk0cMbykMek6THipSXAD5/BKh8H/3JGZTxPgM9Px9WDL0CkM1ORJie48nsWAXQ8kW1YxlknKfIWJs/EBXgoZ6Jf2KMNMYz4FgBJjTGkxR/H67vm/H8eP9ShlyRqfli24c0svy0zLNXgOkNtQJEle/P/MPOv8T3TGZIZIbO7sL7BMON74nkuQqUj4XvnMvwiNCBjO+yev2NVDtZLeX5rvD9lu0zauxW+a6dBvJ8H5Gyfzz3wIBkO57rYECyHeeWF+xW+YcT47Jkdzi4TpT+lPNdIv9Z34fxNOxf0PhO91yw5MuMen56AxLPOtG7W9T63SCQ2k9Uol1so3bVnrog2JTyU57n1bb37n3s5s8Of5RfsaTdSlfuyUAAAAA8dEVYdGNvbW1lbnQAIEltYWdlIGdlbmVyYXRlZCBieSBHTlUgR2hvc3RzY3JpcHQgKGRldmljZT1wbm1yYXcpCvqLFvMAAABKdEVYdHNpZ25hdHVyZQA4NWUxYWU0YTJmYmE3OGVlZDRmZDhmMGFjZjIzNzYwOWU4NGY1NDk2Y2RlMjBiNWQ3NmM5Y2JjMjk4YzRhZWJjJecJ2gAAAABJRU5ErkJggg=="
    />
  </div>
   
  <div
    class="unsafe"
  >
    <img
      onerror="alert('xss');"
      src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAADcAAAA3CAAAAACNsI2aAAAACXBIWXMAAAB5AAAAeQBPsriEAAAB6ElEQVR42rVWO46EMAzNadAcY3vaOQMXoXcXKZehS8NpqNxamw8JxDYra1Zjhgge9jhx/By7bYvtl4Y8Qn+tEjty6WxuQ0KkfOM5wJEeEkT1bsigU+xGQV+QfZ2ned0LAkLnyQ4XV2XB/k+jXdTs8Mc1+UlvQehEt5Fit7hLFsUfqfOk3d1lJ9VO+qN1sFvJm+IScB7s3uo8ZVzC8RrsXjIuqp2n0d+sxFNbHxCw9cF34yn2L5jyJWndIprzRfqLpvw0+6PCh1fjgxpP5NL4VzlYEa6zOYDgzyvk0cMbykMek6THipSXAD5/BKh8H/3JGZTxPgM9Px9WDL0CkM1ORJie48nsWAXQ8kW1YxlknKfIWJs/EBXgoZ6Jf2KMNMYz4FgBJjTGkxR/H67vm/H8eP9ShlyRqfli24c0svy0zLNXgOkNtQJEle/P/MPOv8T3TGZIZIbO7sL7BMON74nkuQqUj4XvnMvwiNCBjO+yev2NVDtZLeX5rvD9lu0zauxW+a6dBvJ8H5Gyfzz3wIBkO57rYECyHeeWF+xW+YcT47Jkdzi4TpT+lPNdIv9Z34fxNOxf0PhO91yw5MuMen56AxLPOtG7W9T63SCQ2k9Uol1so3bVnrog2JTyU57n1bb37n3s5s8Of5RfsaTdSlfuyUAAAAA8dEVYdGNvbW1lbnQAIEltYWdlIGdlbmVyYXRlZCBieSBHTlUgR2hvc3RzY3JpcHQgKGRldmljZT1wbm1yYXcpCvqLFvMAAABKdEVYdHNpZ25hdHVyZQA4NWUxYWU0YTJmYmE3OGVlZDRmZDhmMGFjZjIzNzYwOWU4NGY1NDk2Y2RlMjBiNWQ3NmM5Y2JjMjk4YzRhZWJjJecJ2gAAAABJRU5ErkJggg=="
    />
  </div>
</div>
`;

exports[`Sanitize Html Mixin Special tags test suit Should sanitize and keep name, target, rel attributes in anchor tag 1`] = `
<div>
  <div
    class="safe"
  >
    <a
      href="http://google.com"
      name="moshe"
      rel="noreferrer noopener"
      target="_blank"
    />
  </div>
   
  <div
    class="unsafe"
  >
    <a
      href="http://google.com"
      name="moshe"
      onerror="alert('xss');"
      rel="noreferrer noopener"
      target="_blank"
    />
  </div>
</div>
`;

exports[`Sanitize Html Mixin Special tags test suit Should sanitize img tag with onerror XSS 1`] = `
<div>
  <div
    class="safe"
  >
    <img
      src="x"
    />
  </div>
   
  <div
    class="unsafe"
  >
    <img
      onerror="alert('xss');"
      src="x"
    />
  </div>
</div>
`;

exports[`Sanitize Html Mixin Special tags test suit Should sanitize javascript scheme from anchor tag 1`] = `
<div>
  <div
    class="safe"
  >
    <a />
  </div>
   
  <div
    class="unsafe"
  >
    <a
      href="javascript:alert('xss');"
    />
  </div>
</div>
`;
